instrumentation/opentelemetry-instrumentation-aws_lambda/src/opentelemetry/instrumentation/aws_lambda/__init__.py

```diff
@@ -0,0 +1,257 @@
+
+        Args:
+            lambda_event: user-defined, so it could be anything, but this
+                method counts it being a map with a 'headers' key
+        Returns:
+            A Context with configuration found in the carrier.
+        """
+        parent_context = None
+
+        xray_env_var = os.environ.get(_X_AMZN_TRACE_ID)
+
+        if xray_env_var:
+            parent_context = AwsXRayPropagator().extract(
+                {TRACE_HEADER_KEY: xray_env_var}
+            )
+
+        if (
+            parent_context
+            and get_current_span(parent_context)
+            .get_span_context()
+            .trace_flags.sampled
+        ):
+            return parent_context
+
+        if event_context_extractor:
+            parent_context = event_context_extractor(lambda_event)
+        else:
+            parent_context = _default_event_context_extractor(lambda_event)
+
+        return parent_context
+
+    def _instrumented_lambda_handler_call(call_wrapped, instance, args, kwargs):
+        orig_handler_name = ".".join(
+            [wrapped_module_name, wrapped_function_name]
+        )
+
+        lambda_event = args[0]
+
+        parent_context = _determine_parent_context(lambda_event)
+
+        tracer = get_tracer(__name__, __version__, tracer_provider)
+
+        with tracer.start_as_current_span(
+            name=orig_handler_name,
+            context=parent_context,
+            kind=SpanKind.SERVER,
+        ) as span:
+            if span.is_recording():
+                lambda_context = args[1]
+                # NOTE: The specs mention an exception here, allowing the
+                # `ResourceAttributes.FAAS_ID` attribute to be set as a span
+                # attribute instead of a resource attribute.
+                #
+                # See more:
+                # https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/faas.md#example
+                span.set_attribute(
+                    ResourceAttributes.FAAS_ID,
+                    lambda_context.invoked_function_arn,
+                )
+                span.set_attribute(
+                    SpanAttributes.FAAS_EXECUTION,
+                    lambda_context.aws_request_id,
+                )
+
+            result = call_wrapped(*args, **kwargs)
+
+        _tracer_provider = tracer_provider or get_tracer_provider()
+        if hasattr(_tracer_provider, "force_flush"):
+            # NOTE: force_flush before function quit in case of Lambda freeze.
+            # Assumes we are using the OpenTelemetry SDK implementation of the
+            # `TracerProvider`.
+            _tracer_provider.force_flush()
+        else:
+            logger.error("TracerProvider was missing `force_flush` method. This is necessary in case of a Lambda freeze and would exist in the OTel SDK implementation.")
```

Comment:
If we use `hasattr` like we do here, we can avoid a dependency on the OpenTelemetry SDK `TracerProvider`. However, we obviously miss out on the typed inference we _could_ have here.